File: leetcode_contest_sol/1.py
def modifiedMatrix(matrix):
    """
    :type matrix: List[List[int]]
    :rtype: List[List[int]]
    """

    m = len(matrix)
    n = len(matrix[0])
    
    answer = [ [0]*n for _ in range(m) ]

    max_a = 0
    for j in range(n):
        for i in range(m):
            answer[i][j] = matrix[i][j]
            max_a = max(max_a, matrix[i][j])
        
        for i in range(m):
            if answer[i][j] == -1 :
                answer[i][j] = max_a
    return answer


queries = [
    [[1,2,-1],[4,-1,6],[7,8,9]],
    [[3,-1],[5,2]]
]


for i in queries:
    print(modifiedMatrix(i))

File: leetcode_contest_sol/1.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long

// UNDERSTAND THIS AFTER LEARNING DP

class Solution {
public:
    vector<vector<int>> modifiedMatrix(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();

        vector<vector<int>> answer;
        for(int i=0; i<m; i++){ vector<int > v(0,n);answer.push_back(v); }

        // optimisation 1 : remember the max for a column
        // better idea : traverse column
        // Learning idea :- On interchanging loop, you traverse COLUMN WISE

        // we need to traverse tiint the VERY LAST ELEMENT in column to make sure we get the maximum_elem
        int max_elem = 0;
        for( int j=0; j<n; j++ ){
            for (int i = 0; i < m; i++) { answer[i][j] = matrix[i][j] ; max_elem = max( max_elem, matrix[i][j] );} 
            for (int i = 0; i < m; i++) { if(answer[i][j] == -1)answer[i][j] = max_elem; }
        }
        return answer;

        

    }
};